<a href="https://colab.research.google.com/github/Arhin-Eben/Machine-learning-with-python/blob/master/Logistic_Regression_W3Schools.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Create a function that uses the model's coefficient and intercept values to return a new value
def logit2prob(logr, x):
  log_odds = logr.coef_ * x + logr.intercept_  #create a formula that looks similar to the one from linear regression
  odds = np.exp(log_odds)  #convert the log-odds to odds we must exponentiate the log-odds
  probability = odds / (1 + odds) #convert it to probability by dividing it by 1 plus the odds
  return(probability)

  print(logit2prob(logr, 3.46))

In [ ]:
#Create a function that uses the model's coefficient and intercept values to return a new value
def logit2prob(logr, x):
  log_odds = logr.coef_ * x + logr.intercept_  #create a formula that looks similar to the one from linear regression
  odds = np.exp(log_odds)  #convert the log-odds to odds we must exponentiate the log-odds
  probability = odds / (1 + odds) #convert it to probability by dividing it by 1 plus the odds
  return(probability)

  print(logit2prob(logr, 3.46))

**LOGISTIC REGRESSION**

Logistic regression aims to solve classification problems

It does this by predicting categorical outcomes

unlike linear regression that predicts a continuous outcome

In [13]:
import numpy as np

#Store the independent variables in X
#Store the dependent variable in y

#X represents the size of a tumor in centimeters
X = np.array([3.78, 2.44, 2.09, 0.14, 1.72, 1.65, 4.92, 4.37, 4.96, 4.52, 3.69, 5.88]).reshape(-1,1)

#Note: X has to be reshaped into a column from a row for the LogisticRegression() function to work.
#y represents whether or not the tumor is cancerous (0 for "No", 1 for "Yes").
y = np.array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1])

In [14]:
#We will use a method from the sklearn module
from sklearn import linear_model

#From the sklearn module we will use the LogisticRegression() method
logr = linear_model.LogisticRegression()
logr.fit(X,y)

LogisticRegression()

In [15]:
#predict if tumor is cancerous where the size is 3.46mm:
predicted = logr.predict(np.array([3.46]).reshape(-1,1))
print(predicted)

[0]


We have predicted that a tumor with a size of 3.46mm will not be cancerous

**Coefficient**

In [16]:
import numpy as np
from sklearn import linear_model

#Reshaped for Logistic function
X = np.array([3.78, 2.44, 2.09, 0.14, 1.72, 1.65, 4.92, 4.37, 4.96, 4.52, 3.69, 5.88]).reshape(-1,1)
y = np.array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1])

logr = linear_model.LogisticRegression()
logr.fit(X,y)

log_odds = logr.coef_
odds = np.exp(log_odds)

print(odds)

[[4.03557295]]


This tells us that as the size of a tumor increases by 1mm

bold text the odds of it being a cancerous tumor increases by 4x

**Probability**

In [20]:
#Create a function that uses the model's coefficient and intercept values to return a new value
def logit2prob(logr, x):
  log_odds = logr.coef_ * x + logr.intercept_  #create a formula that looks similar to the one from linear regression
  odds = np.exp(log_odds)  #convert the log-odds to odds we must exponentiate the log-odds
  probability = odds / (1 + odds) #convert it to probability by dividing it by 1 plus the odds
  return(probability)

# Call the function and print the result
print(logit2prob(logr, X))

[[0.60749168]
 [0.19267555]
 [0.12774788]
 [0.00955056]
 [0.08037781]
 [0.0734485 ]
 [0.88362857]
 [0.77901203]
 [0.88924534]
 [0.81293431]
 [0.57718238]
 [0.96664398]]


Results Explained
3.78 0.61 The probability that a tumor with the size 3.78cm is cancerous is 61%.


2.44 0.19 The probability that a tumor with the size 2.44cm is cancerous is 19%.


2.09 0.13 The probability that a tumor with the size 2.09cm is cancerous is 13%.